.github/workflows/test_runner.yaml

```diff
@@ -1,40 +1,82 @@
 name: Tests
 
 on: [push]
+
 env:
   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
 jobs:
-  test_job:
+  install:
+    name: Install and cache dependencies
     runs-on: ubuntu-latest
-    name: Run test suite
+
     steps:
-    - name: Checkout
-      uses: actions/checkout@v1
-    - name: Set up Python 3.7
-      uses: actions/setup-python@v1
-      with:
-        python-version: 3.7
-    - name: Install dependencies
-      run: |
-        python -m pip install --upgrade pip
-        pip install -r requirements.txt
-    - name: Start SQL Server and Presto instances for the tests
-      run: docker-compose up -d mssql presto
-    - name: Run actual tests
-      run: PYTHONPATH=. pytest tests/
-  test-package-build:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: "3.7"
+      - name: Install
+        run: |
+          set -xe
+          python -m pip install --upgrade pip
+          python -m venv venv
+          venv/bin/pip install --progress-bar=off --requirement requirements.txt
+      - name: Cache
+        uses: actions/cache@v2
+        with:
+          path: ${{ github.workspace }}/venv
+          key: venv-${{ github.ref }}-${{ hashFiles('requirements.txt') }}
+
+  check:
+    name: Check formatting
+    needs: install
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: "3.7"
+      - uses: actions/cache@v2
+        with:
+          path: ${{ github.workspace }}/venv
+          key: venv-${{ github.ref }}-${{ hashFiles('requirements.txt') }}
+      - name: Run all pre-commit checks
+        run: |
+          source ${{ github.workspace }}/venv/bin/activate
+          pre-commit run --all-files
```

Comment:
This is pretty neat, but the downside is local tooling can now differ from what's run in CI because dependabot can bump the requirements in requirements.txt but not .pre-commit-config.yaml.  I'm not sure that will actually cause a problem yet, just flagging it!

Reply:
Yeah -- I began to look into getting pre-commit to use the cached virtualenv, but ran out of steam quite quickly!

I think this is a strict improvement on what we have now, but it's something to be aware of.

Reply:
> strict improvement on what we have now

Big +1

I've tried the cached virtualenv thing in the past and also gave up 😬

I'm curious to see how this flow works out 🙂